viabel/optimizers.py

```diff
@@ -0,0 +1,808 @@
+                variational_param_post_conv_history.append(variational_param)
+
+            if sto_process_convergence and j > 200 and t%(2*eval_elbo) ==0:
+                variational_param_post_conv_history_array= np.array(variational_param_post_conv_history)
+                variational_param_post_conv_history_list.append(variational_param_post_conv_history_array)
+                variational_param_post_conv_history_chains = np.stack(variational_param_post_conv_history_list, axis=0)
+                variational_param_post_conv_history_list.pop(0)
+                pmz_size = variational_param_post_conv_history_chains.shape[2]
+                Neff = np.zeros(pmz_size)
+                Rhot = []
+                khat_iterates = []
+                khat_iterates2 = []
+                # compute khat for iterates
+                for k in range(pmz_size):
+                    neff, rho_t_sum, autocov, rho_t = autocorrelation(variational_param_post_conv_history_chains, 0, k)
+                    #mcse_se_combined = monte_carlo_se2(variational_param_history_chains, start_stats,i)
+                    Neff[k] = neff
+                    #mcmc_se2.append(mcse_se_combined)
+                    Rhot.append(rho_t)
+                    khat_i = compute_khat_iterates(variational_param_post_conv_history_chains, 0, k, increasing=True)
+                    khat_iterates.append(khat_i)
+                    khat_i2 = compute_khat_iterates(variational_param_post_conv_history_chains, 0, k, increasing=False)
+                    khat_iterates2.append(khat_i2)
+
+                rhot_array = np.stack(Rhot, axis=0)
+                khat_combined = np.maximum(khat_iterates, khat_iterates2)
+
+
+    if sto_process_convergence:
+
+        optimisation_log['start_avg_mean_iters'] = start_swa_m_iters
+        optimisation_log['start_avg_sigma_iters'] = start_swa_s_iters
+        optimisation_log['r_hat_mean_halfway'] = rhat_mean_halfway
+        optimisation_log['r_hat_sigma_halfway'] = rhat_sigma_halfway
+
+        try:
+            Neff
+        except NameError:
+            pass
+        else:
+            optimisation_log['neff'] = Neff
+            optimisation_log['autocov'] = autocov
+            optimisation_log['rhot'] = rhot_array
+            optimisation_log['start_stats'] = start_stats
+            # optimisation_log['mcmc_se2'] = mcmc_se2_array
+            optimisation_log['khat_iterates_comb'] = khat_combined
+
+    if stopping_rule == 1:
+        start_stats = t - tail_avg_iters
+
+    if stopping_rule ==1:
+        smoothed_opt_param = np.mean(variational_param_history_array[start_stats:,:], axis=0)
+    elif stopping_rule ==2 and sto_process_convergence== True:
+        smoothed_opt_param = np.mean(variational_param_post_conv_history_chains[0, :], axis=0)
+
+    if stopping_rule == 2 and sto_process_convergence == False:
+        smoothed_opt_param = np.mean(variational_param_history_array[start_stats:,:], axis=0)
+
+    if plotting:
```

Comment:
Plotting functionality should be separated out

Reply:
I will try to refactor this in the next commit ...